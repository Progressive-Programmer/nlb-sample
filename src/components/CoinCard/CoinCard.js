import React, { useState } from "react";
import parse from "html-react-parser";
import Typography from "@mui/material/Typography";
import Modal from "@mui/material/Modal";
import Avatar from "@mui/material/Avatar";

import { styled } from "@mui/material/styles";
import Card from "@mui/material/Card";
import CardHeader from "@mui/material/CardHeader";
import CardContent from "@mui/material/CardContent";
import CardActions from "@mui/material/CardActions";
import Collapse from "@mui/material/Collapse";
import IconButton from "@mui/material/IconButton";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";

import Grid from "@mui/material/Unstable_Grid2";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableRow from "@mui/material/TableRow";
import Link from "@mui/material/Link";
import CloseIcon from "@mui/icons-material/Close";

const style = {
  position: "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  overflow: "scroll",
  minWidth: 600,
  maxHeight: 600,
  bgcolor: "background.paper",
  boxShadow: 24,
  p: 4,
};

const ExpandMore = styled((props) => {
  const { expand, ...other } = props;
  return <IconButton {...other} />;
})(({ theme, expand }) => ({
  transform: !expand ? "rotate(0deg)" : "rotate(180deg)",
  marginLeft: "auto",
  transition: theme.transitions.create("transform", {
    duration: theme.transitions.duration.shortest,
  }),
}));

const CoinCard = (props) => {
  const { isLoading, isOpen, data, setIsModalOpen } = props;
  const [expanded, setExpanded] = useState(false);

  const handleClose = () => {
    setIsModalOpen(false);
    setExpanded(false); //closes description accordion as well when box closes
  };

  const handleExpandClick = () => {
    setExpanded(!expanded);
  };

  return (
    <div>
      <Modal
        open={isOpen}
        onClose={handleClose}
        aria-labelledby="modal-modal-title"
        aria-describedby="modal-modal-description"
      >
        <Card sx={style}>
          <CardHeader // Modal Header
            avatar={<Avatar alt={data?.name} src={data?.image?.thumb} />}
            title={data?.name}
            subheader={data?.symbol}
            action={
              <IconButton aria-label="close" onClick={handleClose}>
                <CloseIcon />
              </IconButton>
            }
          />
          <CardContent>
            <Grid
              container
              rowSpacing={1}
              columnSpacing={{ xs: 1, sm: 2, md: 3 }}
            >
              <Grid xs={6}>
                <Table aria-label="a dense table" size="small">
                  <TableBody>
                    <TableRow key={`${data.name}-1`} sx={{ th: { border: 0 } }}>
                      <TableCell component="th" scope="row">
                        <Typography variant="body2" color="text.secondary">
                          <Link href={data?.links?.homepage}>Homepage</Link>
                        </Typography>
                      </TableCell>
                    </TableRow>
                    <TableRow
                      key={`${data.name}-2`}
                      sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                    >
                      <TableCell component="th" scope="row">
                        <Typography variant="body2" color="text.secondary">
                          Market cap in Euro
                        </Typography>
                      </TableCell>
                      <TableCell align="right">
                        {data?.market_cap?.eur ? data.market_cap.eur : "--"}
                      </TableCell>
                    </TableRow>
                  </TableBody>
                </Table>
              </Grid>
              <Grid xs={6}>
                <Table aria-label="a dense table" size="small">
                  <TableBody>
                    <TableRow
                      key={`${data.name}-3`}
                      sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                    >
                      <TableCell component="th" scope="row">
                        <Typography variant="body2" color="text.secondary">
                          Hashing Algorithm
                        </Typography>
                      </TableCell>
                      <TableCell align="right">
                        {data?.hashing_algorithm
                          ? data.hashing_algorithm
                          : "--"}
                      </TableCell>
                    </TableRow>
                    <TableRow
                      key={`${data.name}-4`}
                      sx={{ "&:last-child td, &:last-child th": { border: 0 } }}
                    >
                      <TableCell component="th" scope="row">
                        <Typography variant="body2" color="text.secondary">
                          Genesis Date
                        </Typography>
                      </TableCell>
                      <TableCell align="right">
                        {data?.genesis_date ? data.genesis_date : "--"}
                      </TableCell>
                    </TableRow>
                  </TableBody>
                </Table>
              </Grid>
            </Grid>
          </CardContent>
          {/* Description Accordion  */}
          <CardActions disableSpacing>
            <ExpandMore
              expand={expanded}
              onClick={handleExpandClick}
              aria-expanded={expanded}
              aria-label="show more"
            >
              <ExpandMoreIcon />
            </ExpandMore>
          </CardActions>
          <Collapse in={expanded} timeout="auto" unmountOnExit>
            <CardContent>
              <Typography paragraph>Description:</Typography>
              <Typography paragraph variant="body2" color="text.secondary">
                {parse(data?.description?.en)}
              </Typography>
            </CardContent>
          </Collapse>
        </Card>
      </Modal>
    </div>
  );
};

export default CoinCard;
