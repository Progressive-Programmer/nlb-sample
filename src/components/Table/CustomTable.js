import React, { useEffect, useState } from "react";
import Paper from "@mui/material/Paper";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TablePagination from "@mui/material/TablePagination";
import TableRow from "@mui/material/TableRow";
import Avatar from "@mui/material/Avatar";
import ListItemAvatar from "@mui/material/ListItemAvatar";
import Loader from "../Backdrop/Loader";
import CoinCard from "../CoinCard/CoinCard";

const CustomTable = (props) => {
  const { isLoading, tableData, columns, coinData, onPageChange, onRowClick } =
    props;
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [controller, setController] = useState({
    page: 0,
    rowsPerPage: 10,
  });

  useEffect(() => {
    // fetch table data on first render and
    // then on every page change
    getData();
  }, [controller]);

  const getData = async () => {
    // material ui default page count starts with zero
    // and the first page for api is one
    await onPageChange(controller.page + 1, controller.rowsPerPage);
  };

  const handleChangePage = (event, newPage) => {
    setController({
      ...controller,
      page: newPage,
    });
  };

  const handleChangeRowsPerPage = (event) => {
    setController({
      ...controller,
      rowsPerPage: parseInt(event.target.value, 10),
      page: 0,
    });
  };

  const handleClickRow = (row) => {
    onRowClick(row.id);
    setIsModalOpen(true);
  };

  return (
    <Paper sx={{ width: "90%", overflow: "hidden" }}>
      {isLoading ? <Loader isOpen={isLoading} /> : null}
      <TableContainer sx={{ maxHeight: 650 }}>
        <Table stickyHeader aria-label="sticky table" size="small">
          <TableHead>
            <TableRow>
              {columns &&
                columns.map(
                  (
                    column // set custom column names
                  ) => (
                    <TableCell
                      key={column.id}
                      align={column.align}
                      style={{ minWidth: column.minWidth }}
                    >
                      {column.label}
                    </TableCell>
                  )
                )}
            </TableRow>
          </TableHead>
          <TableBody>
            {tableData
              ? tableData // maping data to each custom row
                  .map((row, i) => {
                    return (
                      <TableRow
                        hover
                        role="checkbox"
                        tabIndex={-1}
                        key={`${row.code}-${i}`}
                        style={{ cursor: "pointer" }}
                        onClick={() => handleClickRow(row)}
                      >
                        {columns &&
                          columns.map((column) => {
                            const value = row[column.id];
                            if (column.id === "image") {
                              // table cell spearation for image type and text
                              return (
                                <TableCell key={column.id} align={column.align}>
                                  <ListItemAvatar>
                                    <Avatar alt={row.name} src={value} />
                                  </ListItemAvatar>
                                </TableCell>
                              );
                            } else {
                              return (
                                <TableCell key={column.id} align={column.align}>
                                  {column.format && typeof value === "number"
                                    ? column.format(value)
                                    : value}
                                </TableCell>
                              );
                            }
                          })}
                      </TableRow>
                    );
                  })
              : "No Data Available"}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={100}
        rowsPerPage={controller.rowsPerPage}
        page={controller.page}
        onPageChange={handleChangePage}
        onRowsPerPageChange={handleChangeRowsPerPage}
      />
      {!isLoading && isModalOpen && (
        <CoinCard
          isLoading={isLoading}
          isOpen={isModalOpen}
          data={coinData ? coinData : {}}
          setIsModalOpen={setIsModalOpen}
        />
      )}
    </Paper>
  );
};

export default CustomTable;
