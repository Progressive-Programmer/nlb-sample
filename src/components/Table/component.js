import React, { useCallback, useEffect, useState } from "react";
import Paper from "@mui/material/Paper";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import TablePagination from "@mui/material/TablePagination";
import TableRow from "@mui/material/TableRow";
import Avatar from "@mui/material/Avatar";
import ListItemAvatar from "@mui/material/ListItemAvatar";
import Loader from "../Backdrop/component";

const StickyHeadTable = (props) => {
  const { isLoading, hasError, currentPage, tableData, columns, onPageChange } = props;
  const [controller, setController] = useState({
    page: 0,
    rowsPerPage: 10,
  });

  useEffect(() => {
    getData();
  }, [controller]);

  const getData = async () => {
    try {
      await onPageChange(controller.page+1, controller.rowsPerPage);
    } catch (error) {
      console.log(error);
    }
  };

  const handleChangePage = (event, newPage) => {
    setController({
      ...controller,
      page: newPage,
    });
  };

  const handleChangeRowsPerPage = (event) => {
    setController({
      ...controller,
      rowsPerPage: parseInt(event.target.value, 10),
      page: 0,
    });
  };

  return (
    <Paper sx={{ width: "90%", overflow: "hidden" }}>
      {isLoading ? <Loader isOpen={isLoading} /> : null}
      <TableContainer sx={{ maxHeight: 650 }}>
        <Table stickyHeader aria-label="sticky table">
          <TableHead>
            <TableRow>
              {columns && columns.map(
                (
                  column // set custom column names
                ) => (
                  <TableCell
                    key={column.id}
                    align={column.align}
                    style={{ minWidth: column.minWidth }}
                  >
                    {column.label}
                  </TableCell>
                )
              )}
            </TableRow>
          </TableHead>
          <TableBody>
            {tableData && tableData // maping data to each custom row
              .map((row, i) => {
                return (
                  <TableRow
                    hover
                    role="checkbox"
                    tabIndex={-1}
                    key={`${row.code}-${i}`}
                  >
                    {columns && columns.map((column) => {
                      const value = row[column.id];
                      if (column.id === "image") {
                        // table cell spearation for image type and text
                        return (
                          <TableCell key={column.id} align={column.align}>
                            <ListItemAvatar>
                              <Avatar alt={row.name} src={value} />
                            </ListItemAvatar>
                          </TableCell>
                        );
                      } else {
                        return (
                          <TableCell key={column.id} align={column.align}>
                            {column.format && typeof value === "number"
                              ? column.format(value)
                              : value}
                          </TableCell>
                        );
                      }
                    })}
                  </TableRow>
                );
              })}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={100}
        rowsPerPage={controller.rowsPerPage}
        page={controller.page}
        onPageChange={handleChangePage}
        onRowsPerPageChange={handleChangeRowsPerPage}
      />
    </Paper>
  );
};

export default StickyHeadTable;
