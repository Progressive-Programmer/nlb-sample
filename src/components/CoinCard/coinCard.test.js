import React from 'react';
import { shallow } from 'enzyme';
import Table from '@mui/material/Table';
import TableRow from '@mui/material/TableRow';
import CardHeader from '@mui/material/CardHeader';
import Collapse from '@mui/material/Collapse';
import CoinCard from './CoinCard';

const setUp = (props = {}) => {
  const wrapper = shallow(<CoinCard {...props} />);
  return wrapper;
};

describe('CoinCard', () => {
  describe('has props', () => {
    let wrapper;
    let props;
    beforeEach(() => {
      props = {
        isOpen: true,
        data: {
          name: 'Test name',
          image: 'image',
          symbol: 'bxy',
          links: {
            homepage: ''
          },
          market_cap: {
            eur: 2
          },
          hashing_algorithm: '',
          genesis_date: '',
          description: {
            en: 'New Description'
          }
        },
        setIsModalOpen: jest.fn()
      };
      wrapper = setUp(props);
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it('should render Card, Grid & Table', () => {
      const component = wrapper;
      expect(component.find(CardHeader)).toHaveLength(1);
      expect(component.find(Table)).toHaveLength(2);
      expect(component.find(TableRow)).toHaveLength(4);
    });
    it('should open Expand when clicked in button ', () => {
      const component = wrapper;
      component.find('[aria-label="show more"]').simulate('click');
      expect(component.find(Collapse).prop('in')).toBe(true);
    });

    it('should close Modal when close is clicked ', () => {
      const setIsModalOpen = props.setIsModalOpen();
      const handleIsModalOpen = jest.spyOn(props, 'setIsModalOpen');
      const component = shallow(<CoinCard {...props} setIsModalOpen={setIsModalOpen} />);

      component.find('[aria-labelledby="modal-modal-title"]').simulate('onClose');
      expect(handleIsModalOpen).toHaveBeenCalled();
    });
  });
});
