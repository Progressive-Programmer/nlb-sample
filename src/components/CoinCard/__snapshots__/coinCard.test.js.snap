// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinCard has props should render correctly 1`] = `
<div>
  <ForwardRef(Modal)
    aria-describedby="modal-modal-description"
    aria-labelledby="modal-modal-title"
    onClose={[Function]}
    open={true}
  >
    <ForwardRef(Card)
      sx={
        Object {
          "bgcolor": "background.paper",
          "boxShadow": 24,
          "left": "50%",
          "maxHeight": 600,
          "minWidth": 600,
          "overflow": "scroll",
          "p": 4,
          "position": "absolute",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        }
      }
    >
      <ForwardRef(CardHeader)
        action={
          <ForwardRef(IconButton)
            aria-label="close"
            onClick={[Function]}
          >
            <Memo />
          </ForwardRef(IconButton)>
        }
        avatar={
          <ForwardRef(Avatar)
            alt="Test name"
          />
        }
        subheader="bxy"
        title="Test name"
      />
      <ForwardRef(CardContent)>
        <ForwardRef(Grid)
          columnSpacing={
            Object {
              "md": 3,
              "sm": 2,
              "xs": 1,
            }
          }
          container={true}
          rowSpacing={1}
        >
          <ForwardRef(Grid)
            xs={6}
          >
            <ForwardRef(Table)
              aria-label="a dense table"
              size="small"
            >
              <ForwardRef(TableBody)>
                <ForwardRef(TableRow)
                  key="Test name-1"
                  sx={
                    Object {
                      "th": Object {
                        "border": 0,
                      },
                    }
                  }
                >
                  <ForwardRef(TableCell)
                    component="th"
                    scope="row"
                  >
                    <ForwardRef(Typography)
                      color="text.secondary"
                      variant="body2"
                    >
                      <ForwardRef(Link)
                        href=""
                      >
                        Homepage
                      </ForwardRef(Link)>
                    </ForwardRef(Typography)>
                  </ForwardRef(TableCell)>
                </ForwardRef(TableRow)>
                <ForwardRef(TableRow)
                  key="Test name-2"
                  sx={
                    Object {
                      "&:last-child td, &:last-child th": Object {
                        "border": 0,
                      },
                    }
                  }
                >
                  <ForwardRef(TableCell)
                    component="th"
                    scope="row"
                  >
                    <ForwardRef(Typography)
                      color="text.secondary"
                      variant="body2"
                    >
                      Market cap in Euro
                    </ForwardRef(Typography)>
                  </ForwardRef(TableCell)>
                  <ForwardRef(TableCell)
                    align="right"
                  >
                    2
                  </ForwardRef(TableCell)>
                </ForwardRef(TableRow)>
              </ForwardRef(TableBody)>
            </ForwardRef(Table)>
          </ForwardRef(Grid)>
          <ForwardRef(Grid)
            xs={6}
          >
            <ForwardRef(Table)
              aria-label="a dense table"
              size="small"
            >
              <ForwardRef(TableBody)>
                <ForwardRef(TableRow)
                  key="Test name-3"
                  sx={
                    Object {
                      "&:last-child td, &:last-child th": Object {
                        "border": 0,
                      },
                    }
                  }
                >
                  <ForwardRef(TableCell)
                    component="th"
                    scope="row"
                  >
                    <ForwardRef(Typography)
                      color="text.secondary"
                      variant="body2"
                    >
                      Hashing Algorithm
                    </ForwardRef(Typography)>
                  </ForwardRef(TableCell)>
                  <ForwardRef(TableCell)
                    align="right"
                  >
                    --
                  </ForwardRef(TableCell)>
                </ForwardRef(TableRow)>
                <ForwardRef(TableRow)
                  key="Test name-4"
                  sx={
                    Object {
                      "&:last-child td, &:last-child th": Object {
                        "border": 0,
                      },
                    }
                  }
                >
                  <ForwardRef(TableCell)
                    component="th"
                    scope="row"
                  >
                    <ForwardRef(Typography)
                      color="text.secondary"
                      variant="body2"
                    >
                      Genesis Date
                    </ForwardRef(Typography)>
                  </ForwardRef(TableCell)>
                  <ForwardRef(TableCell)
                    align="right"
                  >
                    --
                  </ForwardRef(TableCell)>
                </ForwardRef(TableRow)>
              </ForwardRef(TableBody)>
            </ForwardRef(Table)>
          </ForwardRef(Grid)>
        </ForwardRef(Grid)>
      </ForwardRef(CardContent)>
      <ForwardRef(CardActions)
        disableSpacing={true}
      >
        <Styled(Component)
          aria-expanded={false}
          aria-label="show more"
          expand={false}
          onClick={[Function]}
        >
          <Memo(ForwardRef(ExpandMoreIcon)) />
        </Styled(Component)>
      </ForwardRef(CardActions)>
      <ForwardRef(Collapse)
        in={false}
        timeout="auto"
        unmountOnExit={true}
      >
        <ForwardRef(CardContent)>
          <ForwardRef(Typography)
            paragraph={true}
          >
            Description:
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            color="text.secondary"
            paragraph={true}
            variant="body2"
          >
            New Description
          </ForwardRef(Typography)>
        </ForwardRef(CardContent)>
      </ForwardRef(Collapse)>
    </ForwardRef(Card)>
  </ForwardRef(Modal)>
</div>
`;
